xonsh/built_ins.py

```diff
@@ -793,34 +806,37 @@ def run_subproc(cmds, captured=False):
     """
     specs = cmds_to_specs(cmds, captured=captured)
     captured = specs[-1].captured
+    background = specs[-1].background
     procs = []
     proc = pipeline_group = None
+    term_pgid = None
     for spec in specs:
         starttime = time.time()
         proc = spec.run(pipeline_group=pipeline_group)
-        procs.append(proc)
-        if ON_POSIX and pipeline_group is None and \
-           spec.cls is subprocess.Popen:
+        if captured != 'object' and proc.pid and pipeline_group is None:
             pipeline_group = proc.pid
-    if not spec.is_proxy:
+            if update_fg_process_group(pipeline_group, background):
+                term_pgid = pipeline_group
+        procs.append(proc)
+    if not all(x.is_proxy for x in specs):
         add_job({
             'cmds': cmds,
             'pids': [i.pid for i in procs],
             'obj': proc,
-            'bg': spec.background,
+            'bg': background,
         })
     if _should_set_title(captured=captured):
         # set title here to get currently executing command
         pause_call_resume(proc, builtins.__xonsh_shell__.settitle)
     # create command or return if backgrounding.
-    if spec.background:
+    if background:
         return
     if captured == 'hiddenobject':
         command = HiddenCommandPipeline(specs, procs, starttime=starttime,
-                                        captured=captured)
+                                        captured=captured, term_pgid=term_pgid)
```

Comment:
It could not be pulled from its procs or specs. For examples: `ls | cat` and `o = $(ls | cat)` has the same proc and specs but one has `term_pgid` to the pgid of `ls`, one is `None`. Unless we repeat the `if` tests inside `CommandPipeline`:
```
        if proc.pid and pipeline_group is None and not spec.is_proxy and \
                captured != 'object':
            pipeline_group = proc.pid
            if update_fg_process_group(pipeline_group, background):
                term_pgid = pipeline_group

```